### DISCUSSION TITLE
Fixing mesh without affecting the calculation

### DISCUSSION THREAD

**Discussion by souzanha** [idaholab/moose#19315] [Q&A Modules: Solid mechanics] [answered]
Hi!
I have a system with and outward flux on one boundary. This flux nucleates a new phase at that specific boundary and I use the tensor mechanics module to see how the mechanical energy affects the evolution of the interface. 

It could be that I'm doing it incorrectly, but I started without any Dirichlet BCs on the edges and then get a change in my geometry.

But using Dirichlet conditions on the edge affects the simulation results. This makes sense, as I'm keeping the nodes there at constant value 0. Is there a way to fix the mesh in position without me affecting the results?

These are the conditions that I'm using: 
##DISPLACEMENT BOUNDARY CONDITIONS, zero total shear and constant volume
```
  [./bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom_edge'
    value = 0
  [../]
  [./top_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'top_edge'
    value = 0
  [../]
  [./left_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left_edge'
    value = 0
  [../]
  [./right_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'right_edge'
    value = 0
  [../]

##PERIODIC BOUNDARY CONDITIONS, TOP BOTTOM
   [./Periodic]
    [./pbc_c]
      variable = c
      auto_direction = 'y'
    [../]

    [./pbc_w]
      variable = w
      auto_direction = 'y'
    [../]
  [../]

  [./right_w]
     type = ADFluxFunction (my own)
     variable = w
     boundary = right_edge
 [../]

  [./left_w]
     type = NeumannBC
     variable = w
     boundary = left_edge
     value = 0
 [../]

  [./left_c]
     type = NeumannBC
     variable = c
     boundary = left_edge
     value = 0
 [../]

  #total free free energy (chemical + elastic)
  [./total_free_energy]
    type = DerivativeSumMaterial
    f_name = f_tot
    sum_materials = 'f_chem f_elast'
    args = 'c'
    derivative_order=2
  [../]

  [./elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = '157 51 53 0 0 0 171 33 0 0 0 0 0 0 0 38 0 0 49 0 52' #unitless
    fill_method = symmetric21 #general symmetry
  [../]

  [./eigenstrain] #nocomplaints
   type = ComputeVariableEigenstrain
   material_property_names = 'e0_11 e0_22 e0_33'
   #eigen_base = 'e0_11 e0_22 e0_33 0 0 0' #cannot convert to float
   eigen_base = '0.05 0.036 -0.019 0 0 0' #eps0, Vegard’s law coefficients
   prefactor = var_dep #the prefactor is defined in a derivative
   args = 'c'
   eigenstrain_name = eigenstrain
  [../]

  [./var_dependence]
   type = DerivativeParsedMaterial
   function = 'c' #0.5*c^2
   args = 'c'
   output_properties = 'var_dep'
   f_name = var_dep
   enable_jit = true #Enable just-in-time compilation of function expressions for faster evaluation
   derivative_order = 2
  [../]

  [./strain]
   type =  ComputeIncrementalSmallStrain #ComputeSmallStrain #strain, without composition part. eps=1/2*(u*nabla+nablau)
   displacements = 'disp_x disp_y'
   eigenstrain_names = eigenstrain #eigenstrains to be applied in this strain calculation. Substracts the eigenstrain from the total strain = mechanical strain.
  [../]

  [./stress]
   type = ComputeFiniteStrainElasticStress #ComputeLinearElasticStress #sigma=C*mechanical strain = total strain
  [../]
```

**Comment by GiudGiud** [comment]
Hello

I think I'd try pinning a single node first.
Please see how to do this here 
https://mooseframework.inl.gov/modules/tensor_mechanics/tutorials/introduction/step03a.html

Guillaume

**Reply by souzanha** [reply]
I see! Sounds like a very nice approach, I will try it.  Thank you for the suggestion and the link

**Comment by hugary1995** [comment, marked as answer]
This is an issue with your formulation rather than with the tensor mechanics module. You need to first think about what boundary conditions you really want for the control volume. 

If this is something at macroscale, you can apply boundary conditions that make sense on that structure. Depending on the actual problem, pinning a single node may or may not be your best choice. 

If this is an RVE, you may choose traditional KUBC or SUBCs. If you choose SUBC then you need to pin an additional node to remove rigid body modes. Alternatively, you can constrain the RVE to have a prescribed homogenized strain/stress. The global strain system in the tensor mechanics module helps you do that.
